### ISSUE TITLE
This condition is duplicated in another if/elseif branch

### ISSUE THREAD

**Issue by adelf** [kalessil/phpinspectionsea#289]
Some pseudocode:
```
$a = []; // get value from first source
if(!count($a)) {
    $a = [1]; // get it from another source
    if(!count($a)) {
        throw new \Exception();
    }
}
```

Warning on second condition.

**Comment by rentalhost** [comment]
It happen because the same expression is tested twice without check if involved variables are writed. 

To fix I guess that the code between the _start of scope_ (_if then_, on this case) and the _expression_ should never write on a variable used on expression.

By other side, will be hard check conditions like that:

```php
if(mt_rand() > 0.5) {
    if(mt_rand() > 0.5) {
    }
}
```

In all case, despite being an _ugly hack_, is store the expression to a variable. Then you can do it twice without problem:

```php
$rand1 = mt_rand();
if($rand1 >= 0.5) {
    $rand2 = mt_rand();
    if($rand2 >= 0.5) {
    }
}
```

In resume: is a valid false-positive for me, but hard do implements.

**Comment by adelf** [comment]
Second problem hits can be reduced with some white-list for functions like mt_rand, array_rand, etc.

**Comment by rentalhost** [comment]
Not really. Is rare you use `mt_rand()` on conditionals like that (twice).

But you can want to do something like that:

```php
$myClass = new MyClass;

if ($myClass->checkSomething()) {
    $myClass->fix();
    if ($myClass->checkSomething()) {
    }
}
```

**Comment by Fleshgrinder** [comment]
I think that this is not a false positive, on the contrary, it seems as if these things violate the single responsibility principle (first example), or employ the inappropriate intimacy smell (last example).

Does anyone have a real world example for this? I cannot see where I would run into such an issue, and the examples given here are to artificial and constructed.

**Comment by rentalhost** [comment]
I have some real examples, but I am not follow totally the SRP. Then, not is valid too. haha

**Comment by adelf** [comment]
It was a real example from my code. But, I refactored it and it become something like that:
```
$a = []; // get value from first source
if(count($a)) {
    return $a;
}

$a = [1]; // get it from another source
if(count($a)) {
    return $a;
}

 throw new \Exception();
```
But, anyway, I'm not 100% sure that first version was an example of bad code. It's a usual business logic code, when app has main and additional ways to get data. It's not a SRP violation. So, even if this behaviour won't be changed, "Try to refactor this" hint should be added. It's not 100% bad code. It's just a little code smell.

**Comment by rentalhost** [comment]
Don't is needly a question of good or bad code, but how to implements
something like that, without discard good cases that works currently.

**Comment by Fleshgrinder** [comment]
I also wouldn’t categorize it as bad code, the multiple return example of yours is something that other people would consider non-optimal compared to the initial example.

I on the other hand would argue that your examples are not generic enough. The question I am asking myself is, why are the sources not throwing exceptions if the decorator does? The whole things looks very much like the repository pattern with a chain. Hence, the first thing we need would be an interface that all of them share.

```php
interface Repo {
  function find(): ?array;
  /** @throws NotFoundException */
  function get(): array;
}
```

Next we need the two sources.

```php
final class Source1 implements Repo {
  public function find(): ? array { /*...*/ }
  /** @throws NotFoundException */
  public function get(): array { /*...*/ }
}

final class Source2 implements Repo {
  public function find(): ? array { /*...*/ }
  /** @throws NotFoundException */
  public function get(): array { /*...*/ }
}
```

And finally we create a decorator that orchestrates between them.

```php
final class Orchestrator implements Repo {
  private $a;
  private $b;
  public function __construct(Repo $a, Repo $b) {
    $this->a = $a;
    $this->b = $b;
  }
  public function find(): ?array {
    $result = $this->a->find();
    if (!$result) {
      $result = $this->b->find();
    }
    return $result;
  }
  /** @throws NotFoundException */
  public function get(): array {
    $result = $this->find();
    if (!$result) {
      throw new NotFoundException;
    }
    return $result;
  }
}
```

Now we have no violations of any principles, and highly generic code that we can reuse in many ways.

**Comment by Alexgmin** [comment]
Even the [basic FizzBuzz](https://gist.github.com/rbl00/850432) gets a warning.
```php
for ($i = 1; $i <= 100; ++$i) {
    if ($i % 3 == 0 && $i % 5 == 0) {
        echo 'FizzBuzz<br />';
    } elseif ($i % 3 == 0) {
        echo 'Fizz<br />';
    } elseif ($i % 5 == 0) {
        echo 'Buzz<br />';
    } else {
        echo $i . '<br />';
    }
}
```

**Comment by kalessil** [comment]
Patterns from @adelf and @Alexgmin have been reported multiple times, and I declining them all the time: it obviously code needs refactoring.

Case 1: candidate for yielding/iterator over sources.
Case 2: has maintainability and extensibility issue, I'd do something like this (perhaps something better, but limited in time):

```php
$strategies = [
    function($parameter)        { return ($parameter % 3 === 0 && $parameter % 5 === 0) ? 'FizzBuzz' : false; },
    function($parameterFizz)    { return $parameterFizz % 3 === 0 ? 'Fizz' : false; },
    function($parameterBuzz)    { return $parameterBuzz % 5 === 0 ? 'Buzz' : false; }
    /* keep extyending here with low risk of introducing *new* bugs */
];

for ($argument = 1; $argument <= 100; ++$argument) {
    $result = $argument;
    foreach ($strategies as $callback) {
        $strategyResult = $callback($argument);
        if (false !== $strategyResult) {
            $result = $strategyResult;
            break;
        }
    }
    echo $result . PHP_EOL;
}
```

**Comment by kalessil** [comment]
Fixed #550, coming into 2.3.12
